### PR TITLE
PLAT-35655: Add ui/Slider and adopt in moonstone

### PR DESCRIPTION
**Pull request by ryanjduffy**
### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)

* Adds `ui/Slider` component as a ground-up rewrite using `ui/Touchable` for drag support
* Major refactor of `moonstone/Slider`, `moonstone/IncrementSlider`, and `moonstone/VideoPlayer`'s MediaSlider to adopt the new base component
* Rewrite `moonstone/Slider` and `moonstone/IncrementSlider` tests to validate the moonstone-specific behaviors

Enact-DCO-1.0-Signed-off-by: Ryan Duffy (ryan.duffy@lge.com)

### CODE REVIEW

File: packages/moonstone/VideoPlayer/VideoPlayer.js [hunk 1]
```diff
@@ -1694,8 +1694,7 @@ const VideoPlayerBase = class extends React.Component {
 		Spotlight.setPointerMode(false);
 
 		if (this.focusDefaultMediaControl()) {
-			ev.preventDefault();
-			ev.stopPropagation();
+			ev.nativeEvent.stopImmediatePropagation();
```
Comment (viodragon2) [inline packages/moonstone/VideoPlayer/VideoPlayer.js]: `ev` is not a native event and will throw an error.

File: packages/moonstone/Slider/utils.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import {calcPercent} from '@enact/ui/Slider/utils';
+
+const isNumber = (n) => typeof n === 'number' && !isNaN(n);
+
+const calcStep = (knobStep, step) => {
+	let s;
+
+	if (isNumber(knobStep)) {
+		s = knobStep;
+	} else if (isNumber(knobStep)) {
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply packages/moonstone/Slider/utils.js]: Good catch. It should be `isNumber(step)`. `knobStep` could be undefined or null or 0 so a straight falsey check isn't possible. We could also test with `!= null` which should be safe.

File: packages/moonstone/Slider/SliderBehaviorDecorator.js [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+	// FIXME: This is a compromise to maintain a single decorator for Slider and IncrementSlider
+	// that handles both a consolidated focus state and spotlight directional event mgmt. When this
+	// is unset (for Slider), this decorator will listen to onKeyDown and fire spotlight events.
+	// When set (for IncrementSlider), it specifies the event that is passed down to trigger
+	// spotlight events and also doesn't remove the spotlight directional callbacks from the props
+	// so the Wrapped component can fire them manually or use the callback for the default behavior.
+	emitSpotlightEvents: null
+};
+
+// Adds moonstone-specific slider behaviors
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply packages/moonstone/Slider/SliderBehaviorDecorator.js]: Yes it is. I didn't use JSdoc because the information is never intended to be consumed outside of the source. Consider it inline docs on the requirements for the component.

File: packages/moonstone/Slider/utils.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+const emitChange = (direction) => adaptEvent(
+	(ev, {knobStep, max, min, step, value}) => ({
+		value: clamp(min, max, value + (calcStep(knobStep, step) * direction)),
+		proportion: calcPercent(min, max, value)
+	}),
+	forward('onChange')
+);
+
+const isActive = (ev, props) => {
+	return props.active || props.activateOnFocus || props.detachedKnob;
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply packages/moonstone/Slider/utils.js]: Correct. Good catch!

File: packages/moonstone/VideoPlayer/MediaSliderDecorator.js [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+import {calcPercent} from '@enact/ui/Slider/utils';
+import PropTypes from 'prop-types';
+import React from 'react';
+
+import MediaKnob from './MediaKnob';
+
+// decrements the MediaKnob position if we're tracking
+const decrement = (state) => {
+	if (state.tracking && state.x > 0) {
+		const x = Math.max(0, state.x - 0.05);
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply packages/moonstone/VideoPlayer/MediaSliderDecorator.js]: I think it's irrelevant in this case but also not feasible to do given the current encapsulation. MediaSliderDecorator wraps around MediaSlider but MediaSlider is the component setting `knobStep`. In practice, that setting isn't used because `knobStep` is only used when activating the knob and then pressing 5-way controls which isn't supported by MediaSlider.

Going to remove the prop from MediaSlider and keep this hard coded value for now.

File: packages/moonstone/VideoPlayer/VideoPlayer.js (hunk 1 above)
Comment (ryanjduffy) [reply packages/moonstone/VideoPlayer/VideoPlayer.js]: You are right! It used to receive the keyboard event but doesn't any longer. I'll need to revisit this functionality.

File: packages/ui/Slider/Slider.js [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+		/**
+		 * The amount to increment or decrement the position of the knob via 5-way controls.
+		 *
+		 * If not specified, `step` is used for the default value.
+		 *
+		 * @type {Number}
+		 * @default 1
+		 * @public
+		 */
+		knobStep: PropTypes.number,
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply packages/ui/Slider/Slider.js]: I moved this prop to `moonstone/Slider` since it isn't currently used by `ui/Slider`

**Comment by ryanjduffy** [comment]
Thanks for the great, detailed feedback. I've made a round of changes and addressed a couple of your comments below. The rest are forthcoming!

> moonstone/Slider positioned to values between steps resulting in decimal point values while dragging. This doesn't seem to happen in IncrementSlider. This is possibly due to no default value assigned for step in Slider. Any specific reason to not have default step value?

I did that intentionally but I think that stepped changes are a better default.

> moonstone/IncrementalSlider does not retain focus while dragging if mouse is poistioned away from the knob.

Not seeing the same behavior locally :/

> moonstone/VideoPlayer's MediaSlider knob color is wrong: the knob color should be white.

Not seeing this locally either

**Comment by ryanjduffy** [comment]
> Let's add a test for handle.adaptEvent.

Done!

> moonstone/VideoPlayer's MediaSlider does not position correctly: It's positioned a bit lower than before. It needs to have a space between Media controls and the slider.

This appears the same to me between develop and this branch. I did a screen shot test as well as comparing dev tools positioning but please share some details if you're still seeing this.

> moonstone/VideoPlayer tooltip with thumbnailSrc does not change sides when it passes middle point. Play time tooltip works okay.

Fixed!

> moonstone/VideoPlayer throws error when spotlight changes from slider to media controls via 5-way

Fixed!

> Should we create tests for ui/Slider components?

Perhaps but I'm not sure what would be a meaningful test. There would definitely be opportunity for UI tests to validate the drag behaviors. I'll think on this a bit more.

File: packages/moonstone/VideoPlayer/FeedbackTooltip.js [hunk 7]
```diff
@@ -138,9 +171,12 @@ const FeedbackTooltipBase = kind({
 	},
 
 	render: ({children, noFeedback, playbackState, playbackRate, thumbnailComponent, ...rest}) => {
-		delete rest.visible;
+		delete rest.duration;
+		delete rest.formatter;
 		delete rest.thumbnailDeactivated;
 		delete rest.thumbnailSrc;
+		delete rest.visible;
```
Comment (viodragon2) [inline packages/moonstone/VideoPlayer/FeedbackTooltip.js]: We should delete `orientation` and `proportion` as well as other props that are passed from the render function of `ui/Slider`

File: packages/sampler/stories/moonstone-stories/IncrementSlider.js [hunk 8]
```diff
@@ -11,29 +11,44 @@ import {mergeComponentMetadata} from '../../src/utils/propTables';
...
 				noFill={boolean('noFill', false)}
 				onChange={action('onChange')}
 				orientation={select('orientation', ['horizontal', 'vertical'], 'horizontal')}
-				step={number('step', IncrementSliderBase.defaultProps.step)}
+				step={number('step', 1)}
 				tooltip={nullify(boolean('tooltip', false))}
-				tooltipAsPercent={nullify(boolean('tooltipAsPercent', false))}
-				tooltipForceSide={nullify(boolean('tooltipForceSide', false))}
-				tooltipSide={select('tooltipSide', ['before', 'after'], 'after')}
+				percent={nullify(boolean('percent', false))}
```
Comment (viodragon2) [inline packages/sampler/stories/moonstone-stories/IncrementSlider.js]: nit: sort by alphabetical order

File: packages/sampler/stories/moonstone-stories/Slider.js [hunk 9]
```diff
@@ -11,30 +11,44 @@ import {mergeComponentMetadata, removeProps} from '../../src/utils/propTables';
...
 				orientation={select('orientation', ['horizontal', 'vertical'], 'horizontal')}
-				step={number('step', SliderBase.defaultProps.step)}
+				step={number('step', 1)}
 				tooltip={nullify(boolean('tooltip', false))}
-				tooltipAsPercent={nullify(boolean('tooltipAsPercent', false))}
-				tooltipForceSide={nullify(boolean('tooltipForceSide', false))}
-				tooltipSide={select('tooltipSide', ['before', 'after'], 'after')}
+				percent={nullify(boolean('percent', false))}
+				side={nullify(select('side', ['before', 'after']))}
+				forceSide={nullify(boolean('forceSide', false))}
```
Comment (viodragon2) [inline packages/sampler/stories/moonstone-stories/Slider.js]: nit: alphabetical order

File: packages/moonstone/VideoPlayer/FeedbackTooltip.js [hunk 10]
```diff
@@ -165,6 +201,8 @@ const FeedbackTooltip = onlyUpdateForKeys(
 	)
 );
 
+FeedbackTooltip.defaultSlot = 'tooltip';
```
Comment (viodragon2) [inline packages/moonstone/VideoPlayer/FeedbackTooltip.js]: doesn't seem like this is getting picked up by `Slottable`

File: packages/ui/Slider/Slider.js [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+/**
+ * Adds touch and drag support to a [SliderBase]{@link ui/Slider.SliderBase}.
+ *
+ * @hoc
+ * @memberof ui/Slider
+ * @mixes ui/Touchable.Touchable
+ * @public
+ */
+const SliderDecorator = compose(
+	PositionDecorator,
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply packages/ui/Slider/Slider.js]: Hmm ... probably. Unfortunately, we need `Changeable` above `moonstone/Slider` because it mutates the value for 5-way changes. So, the consequence of adding is that we either:

* have an extra layer that isn't adding value, or
* we have to recompose the HOCs in moonstone to avoid the above.

Will experiment with different HOC ordering to avoid both of the above but will likely fall back to the second option.

File: packages/moonstone/IncrementSlider/IncrementSlider.js [hunk 12]
```diff
@@ -650,14 +534,17 @@ const IncrementSliderBase = kind({
 
 const IncrementSliderDecorator = compose(
 	Pure,
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply packages/moonstone/IncrementSlider/IncrementSlider.js]: Just forgot it!

File: packages/moonstone/ProgressBar/ProgressBarTooltip.js [hunk 13]
```diff
@@ -142,5 +176,10 @@ const ProgressBarTooltipBase = kind({
 	}
 });
 
+ProgressBarTooltipBase.defaultSlot = 'tooltip';
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply packages/moonstone/ProgressBar/ProgressBarTooltip.js]: It works in Slider but not in ProgressBar (yet) because it lacks `Slottable` support. I'm adding that support in #1573 which standardizes tooltip-related props across all three components.

File: packages/ui/Slider/Slider.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply packages/ui/Slider/Slider.js]: I think I've found a way to include `Changeable` in `ui/SliderDecorator` and then compose `ui/SliderDecorator` in `moonstone/Slider` and `moonstone/IncrementSlider` with minimal extra layers.

File: packages/ui/Slider/PositionDecorator.js [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+			if (step) {
+				const delta = (value - min) % step;
+				if (delta < step / 2) {
+					value -= delta;
+				} else {
+					value += step - delta;
+				}
+			}
+
+			forward('onChange', {
```
Comment (viodragon2) [inline packages/ui/Slider/PositionDecorator.js]: Should we fire when only proportion changes? I'd assumed it fires only when "value" changes. With current implementation, it fires for every little mouse movement, and I'm afraid it could cause perf issues.

Another thing we can think about is to add a threshold for throttling. If steps are small enough that causes to fire multiple onChange event, it could cause perf issues. This can be an additive change in the future, so I don't mind not having that feature for now, but it's something to think about.

Comment (ryanjduffy) [reply packages/ui/Slider/PositionDecorator.js]: Added a bit of code to prevent emitting `onChange` when the value hadn't changed. I also corrected a bug that would emit an invalid `proportion` when `step` is used and dragging between step points.

Regarding throttling, I'm not sure we can effectively implement that on the framework. I think the onus lies on the application to either set a `step` to limit the amount of changes or throttle the effects of the `onChange` if the app wants a smooth slider but the side effects are expensive.

**Comment by ryanjduffy** [comment]
The duplicate event issue was due to the HOC refactoring. Glad you caught it! I had used `ui/Slider` instead of `SliderBase` so there were dual drag handlers in effect. After fixing that I discovered there was no good way to reuse `ui/Slider.SliderDecorator` because I needed to inject the moonstone slider hoc, `SliderBehaviorDecorator` between `Changeable` and `PositionDecorator`. Rather than duplicate a the HOCs from ui, I elected to use `ui/Slider` directly and add an extra `Changeable` on top of it all. It's not ideal but it's the simplest solution. We can revisit later if we want to tweak for performance.

**Comment by viodragon2** [comment]
LGTM

File: packages/moonstone/Slider/utils.js [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+};
+
+const isDecrement = ({keyCode}, {orientation}) => {
+	return orientation === 'vertical' ? is('down', keyCode) : is('left', keyCode);
+};
+
+const emitChange = (direction) => adaptEvent(
+	(ev, {knobStep, max, min, step, value = min}) => ({
+		value: clamp(min, max, value + (calcStep(knobStep, step) * direction)),
+		proportion: calcProportion(min, max, value),
```
Comment (webOS101) [inline packages/moonstone/Slider/utils.js]: Proportion is calculated with the previous value, not the calculated value.

**Comment by ryanjduffy** [comment]
Good catch on `IncrementSlider`. I had a crufty HOC in place there causing multiple drag handlers to be in play and removing it did the trick. Also addressed your other feedback.
